### PR TITLE
Add ContractDoesNotImplementInterface error.

### PR DESCRIPTION
**Pull request by sofiafaro-da**
For when fetching or exercising by interface generically, and the fetched contract doesn't implement the interface. Right now this takes precedence over type errors, but it shouldn't.

Part of #10810 and #11703.

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/transaction/src/main/scala/com/daml/lf/Error.scala [hunk 1]
```diff
@@ -89,6 +89,15 @@ object Error {
       actual: TypeConName,
   ) extends Error
 
+  /** We tried to fetch / exercise a contract by interface, but
+    * the contract does not implement this interface.
+    */
+  final case class ContractDoesntImplementInterface(
```
Comment (cocreature) [inline daml-lf/transaction/src/main/scala/com/daml/lf/Error.scala:95]: bikeshed: Should it be `ContractDoesNotImplementInterface`?

File: daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/InterfacesTest.scala [hunk 2]
```diff
@@ -255,7 +281,16 @@ class InterfacesTest
     "be unable to exercise T1 (disguised as T2) by interface I2 via 'exercise by interface'" in {
       val command =
         ExerciseByInterfaceCommand(idI2, idT2, cid1, "C2", ValueRecord(None, ImmArray.empty))
-      run(command) shouldBe a[Left[_, _]]
+      run(command) match {
+        case Left(Error.Interpretation(err, _)) =>
```
Comment (cocreature) [inline daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/InterfacesTest.scala:285]: you might want to use `inside` to make the assertion a bit nicer.

File: daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/InterfacesTest.scala [hunk 3]
```diff
@@ -283,7 +318,16 @@ class InterfacesTest
     }
     "be unable to fetch T1 (disguised as T2) via interface I2" in {
       val command = FetchByInterfaceCommand(idI2, idT2, cid1)
-      run(command) shouldBe a[Left[_, _]]
+      run(command) match {
```
Comment (cocreature) [inline daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/InterfacesTest.scala:321]: same here.

File: daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/InterfacesTest.scala (hunk 2 above)
Comment (cocreature) [reply daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/InterfacesTest.scala:285]: and for most of the other cases as well

**Review by cocreature — APPROVED** [review]
Thanks!

**Review by remyhaemmerle-da — APPROVED** [review]

File: daml-lf/transaction/src/main/scala/com/daml/lf/Error.scala (hunk 1 above)
Comment (sofiafaro-da) [reply daml-lf/transaction/src/main/scala/com/daml/lf/Error.scala]: done!

File: daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/InterfacesTest.scala (hunk 2 above)
Comment (sofiafaro-da) [reply daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/InterfacesTest.scala]: thanks for the suggestion!

**Pull request merged**